File: src/nnue.py
import numpy as np

# ---- Network hyperparameters (from your training config) ----
HIDDEN_SIZE: int = 128
QA: np.int16 = np.int16(255)
QB: np.int16 = np.int16(64)
SCALE: np.int32 = np.int32(400)

# Piece kind indices
PIECE_CHARS = "PNBRQKpnbrqk"
#              0=pawn, 1=knight, 2=bishop, 3=rook, 4=queen, 5=king
#              colour = idx // 6  (0=white, 1=black)
#              kind   = idx %  6


def screlu(x: np.int16) -> np.int32:
    """
    Square Clipped ReLU - Activation Function.
    Range is 0.0 .. 1.0 (in other words, 0 to QA*QA quantized).
    """
    y = np.int32(x).clip(0, np.int32(QA))
    return y * y


# ---------------------------------------------------------------------------
# Piece / Position
# ---------------------------------------------------------------------------

class Piece:
    """
    One piece on the board, in absolute (non-perspective-flipped) terms.

    colour : 0 = white, 1 = black
    kind   : 0=pawn 1=knight 2=bishop 3=rook 4=queen 5=king
    square : 0..63  (a1=0, h8=63)
    """

    def __init__(self, colour: int, kind: int, square: int):
        self.colour = colour
        self.kind   = kind
        self.square = square


class Position:
    """
    A board position: a list of pieces and the side to move.

    stm : 0 = white to move, 1 = black to move
    """

    def __init__(self, pieces: list[Piece], stm: int):
        self.pieces = pieces
        self.stm    = stm


# ---------------------------------------------------------------------------
# FEN parser
# ---------------------------------------------------------------------------

def parse_fen(fen: str) -> Position:
    """
    Parses a FEN string into a Position.
    Mirrors the Rust parse_fen function exactly.
    Raises ValueError on malformed input.
    """
    parts = fen.split()
    if not parts:
        raise ValueError("empty FEN")

    board_str = parts[0]

    # Side to move
    stm = 1 if (len(parts) > 1 and parts[1] == "b") else 0

    ranks = board_str.split("/")
    if len(ranks) != 8:
        raise ValueError(f"FEN board must have 8 ranks, found {len(ranks)}")

    pieces: list[Piece] = []

    # FEN lists ranks from 8 down to 1; rank 1 is rank index 0.
    for i, rank_str in enumerate(ranks):
        rank = 7 - i
        file = 0

        for ch in rank_str:
            if ch.isdigit():
                file += int(ch)
            else:
                idx = PIECE_CHARS.find(ch)
                if idx == -1:
                    raise ValueError(f"invalid piece character '{ch}'")
                if file >= 8:
                    raise ValueError(f"rank '{rank_str}' overflows 8 files")

                pieces.append(Piece(
                    colour = idx // 6,
                    kind   = idx %  6,
                    square = rank * 8 + file,
                ))
                file += 1

        if file != 8:
            raise ValueError(f"rank '{rank_str}' does not sum to 8 files")

    return Position(pieces, stm)


# ---------------------------------------------------------------------------
# Accumulator / Network
# ---------------------------------------------------------------------------

class Accumulator:
    """
    A column of the feature-weights matrix.
    Equivalent to repr(C, align(64)) — alignment is handled by NumPy internally.
    """

    def __init__(self, vals: np.ndarray):
        # vals: [i16; HIDDEN_SIZE]
        assert vals.dtype == np.int16
        assert vals.shape == (HIDDEN_SIZE,)
        self.vals = vals.copy()

    def add_feature(self, feature_idx: int, net: "Network"):
        """Adds a feature column from the network's feature weights."""
        self.vals += net.feature_weights[feature_idx].vals

    def remove_feature(self, feature_idx: int, net: "Network"):
        """Removes a feature column from the network's feature weights."""
        self.vals -= net.feature_weights[feature_idx].vals


class Network:
    """
    Quantised NNUE network.
    Mirrors the repr(C) Rust struct — fields are loaded in declaration order
    from the binary checkpoint file that bullet produces.

    Layout:
        feature_weights : [Accumulator; 768]  →  shape (768, HIDDEN_SIZE) i16
        feature_bias    : Accumulator         →  shape (HIDDEN_SIZE,)      i16
        output_weights  : [i16; 2*HIDDEN_SIZE]→  shape (2*HIDDEN_SIZE,)    i16
        output_bias     : i16                 →  scalar                    i16
    """

    def __init__(
        self,
        feature_weights: np.ndarray,   # (768, HIDDEN_SIZE) int16
        feature_bias: np.ndarray,      # (HIDDEN_SIZE,)     int16
        output_weights: np.ndarray,    # (2*HIDDEN_SIZE,)   int16
        output_bias: np.int16,
    ):
        assert feature_weights.shape == (768, HIDDEN_SIZE)
        assert feature_bias.shape    == (HIDDEN_SIZE,)
        assert output_weights.shape  == (2 * HIDDEN_SIZE,)

        # Wrap each row of feature_weights as an Accumulator
        self.feature_weights: list[Accumulator] = [
            Accumulator(feature_weights[i]) for i in range(768)
        ]
        self.feature_bias   = Accumulator(feature_bias)
        self.output_weights = output_weights.astype(np.int16)
        self.output_bias    = np.int16(output_bias)

    @classmethod
    def load(cls, path: str) -> "Network":
        """
        Load a quantised bullet checkpoint (.bin) directly from disk.
        Mirrors Rust's:
            std::mem::transmute(*include_bytes!("../checkpoints/quantised_2.bin"))
        Because the Rust struct is repr(C), the binary layout is well-defined
        and we can read it field-by-field in declaration order.
        """
        with open(path, "rb") as f:
            data = f.read()

        offset = 0

        # feature_weights: [Accumulator; 768] = 768 * HIDDEN_SIZE i16 values
        count = 768 * HIDDEN_SIZE
        feature_weights = np.frombuffer(data, dtype=np.int16, count=count, offset=offset)
        feature_weights = feature_weights.reshape((768, HIDDEN_SIZE)).copy()
        offset += count * 2  # 2 bytes per i16

        # feature_bias: Accumulator = HIDDEN_SIZE i16 values
        feature_bias = np.frombuffer(data, dtype=np.int16, count=HIDDEN_SIZE, offset=offset).copy()
        offset += HIDDEN_SIZE * 2

        # output_weights: [i16; 2 * HIDDEN_SIZE]
        count = 2 * HIDDEN_SIZE
        output_weights = np.frombuffer(data, dtype=np.int16, count=count, offset=offset).copy()
        offset += count * 2

        # output_bias: i16 (single scalar)
        output_bias = np.frombuffer(data, dtype=np.int16, count=1, offset=offset)[0]

        return cls(feature_weights, feature_bias, output_weights, output_bias)

    def evaluate(self, us: Accumulator, them: Accumulator) -> np.int32:
        """
        Calculates the output of the network from the two perspective accumulators.

        us   = accumulator from the perspective of the side to move
        them = accumulator from the opponent's perspective
        """
        output = np.int32(0)

        # First half of output_weights corresponds to `us`
        for inp, weight in zip(us.vals, self.output_weights[:HIDDEN_SIZE]):
            output += screlu(inp) * np.int32(weight)

        # Second half corresponds to `them`
        for inp, weight in zip(them.vals, self.output_weights[HIDDEN_SIZE:]):
            output += screlu(inp) * np.int32(weight)

        output //= np.int32(QA)
        output  += np.int32(self.output_bias)
        output  *= SCALE
        output //= np.int32(QA) * np.int32(QB)

        return output

    def make_accumulator(self) -> Accumulator:
        """
        Creates a fresh accumulator initialised to the feature bias.
        Equivalent to Accumulator::new(&net) in Rust.
        """
        return Accumulator(self.feature_bias.vals.copy())


# ---------------------------------------------------------------------------
# Accumulator builder  (Chess768 feature layout)
# ---------------------------------------------------------------------------

def build_accumulators(net: Network, pos: Position) -> tuple[Accumulator, Accumulator]:
    """
    Builds the two perspective accumulators (us = side to move, them = other
    side) for a position, using Chess768's feature layout:

        feature = colour * 384 + piece_kind * 64 + square

    mirroring the square (^ 56) and flipping colour to get the
    opposite-perspective feature.
    """
    us   = net.make_accumulator()
    them = net.make_accumulator()

    for p in pos.pieces:
        white_feat = p.colour * 384 + p.kind * 64 + p.square
        black_feat = (1 - p.colour) * 384 + p.kind * 64 + (p.square ^ 56)

        if pos.stm == 0:          # white to move
            us_feat, them_feat = white_feat, black_feat
        else:                     # black to move
            us_feat, them_feat = black_feat, white_feat

        us.add_feature(us_feat, net)
        them.add_feature(them_feat, net)

    return us, them


# ---------------------------------------------------------------------------
# Top-level FEN evaluation API
# ---------------------------------------------------------------------------

def evaluate_fen(fen: str) -> int:
    """
    Parses `fen`, builds accumulators, and runs the forward pass.
    Returns the evaluation in centipawns, relative to the side to move.
    Mirrors evaluate_fen / evaluate_fen_api in Rust.
    """
    net = Network.load("../checkpoints/quantised_2.bin")
    pos = parse_fen(fen)
    us, them = build_accumulators(net, pos)
    return int(net.evaluate(us, them))

# this one is so we don't have to keep reloading the network parameters
def evaluate_fen(net: Network, fen: str) -> int:
    pos = parse_fen(fen)
    us, them = build_accumulators(net, pos)
    return int(net.evaluate(us, them))

# ---------------------------------------------------------------------------
# Example usage
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    # net = Network.load("checkpoints/quantised_2.bin")

    # Starting position
    fen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
    score = evaluate_fen(fen)
    print(f"Starting position: {score} cp (relative to side to move)")

    # After 1.e4
    fen2 = "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e3 0 1"
    score2 = evaluate_fen(fen2)
    print(f"After 1.e4:        {score2} cp (relative to side to move)")

File: nnue_test/nnue_test/src/nnue.rs
//! Minimal FEN -> Chess768 -> NNUE eval tool, built around bullet's
//! quantised inference format.
//!
//! Usage:
//!     cargo run --release -- "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
//!
//! Feature encoding was cross-checked against bullet's own data format
//! (jw1912/bulletformat, src/chess.rs) rather than assumed:
//!   - piece nibble = (colour << 3) | piece_kind
//!   - piece_kind:   0=Pawn 1=Knight 2=Bishop 3=Rook 4=Queen 5=King
//!   - square:       rank * 8 + file, a1 = 0 .. h8 = 63 (LERF)
//! Chess768's 768 = 2 colours * 6 piece kinds * 64 squares comes directly
//! from that same (colour, piece, square) triple, so the feature index is
//! `colour * 384 + piece_kind * 64 + square`, with the standard perspective-net
//! trick of mirroring the square (^ 56) and flipping colour for the
//! not-side-to-move accumulator.
//!
//! If your training run used a different piece ordering than the above,
//! everything downstream (PIECE_CHARS mapping in `parse_fen`) needs updating
//! to match, or you'll get a network that runs but evaluates nonsense.

// ---- Network hyperparameters (from your training config) ----
const HIDDEN_SIZE: usize = 128;
const QA: i16 = 255;
const QB: i16 = 64;
const SCALE: i32 = 400;

/// Loaded at compile time from the checkpoint bullet produced.
/// Path is relative to *this source file*.
static NNUE: Network = unsafe {
    std::mem::transmute(*include_bytes!("../checkpoints/quantised_2.bin"))
};

#[inline]
/// Square Clipped ReLU - Activation Function.
/// Range is 0.0 .. 1.0 (in other words, 0 to QA*QA quantized).
fn screlu(x: i16) -> i32 {
    let y = i32::from(x).clamp(0, i32::from(QA));
    y * y
}

/// This is the quantised format that bullet outputs.
#[repr(C)] // this means represent it like C does (fields in order)
pub struct Network {
    feature_weights: [Accumulator; 768],
    feature_bias: Accumulator,
    output_weights: [i16; 2 * HIDDEN_SIZE],
    output_bias: i16,
}

impl Network {
    /// Calculates the output of the network from the two perspective accumulators.
    pub fn evaluate(&self, us: &Accumulator, them: &Accumulator) -> i32 {
        let mut output = 0;

        for (&input, &weight) in us.vals.iter().zip(&self.output_weights[..HIDDEN_SIZE]) {
            output += screlu(input) * i32::from(weight);
        }

        for (&input, &weight) in them.vals.iter().zip(&self.output_weights[HIDDEN_SIZE..]) {
            output += screlu(input) * i32::from(weight);
        }

        output /= i32::from(QA);
        output += i32::from(self.output_bias);
        output *= SCALE;
        output /= i32::from(QA) * i32::from(QB);

        output
    }
}

/// A column of the feature-weights matrix.
#[derive(Clone, Copy)]
#[repr(C, align(64))]
pub struct Accumulator {
    vals: [i16; HIDDEN_SIZE],
}

impl Accumulator {
    pub fn new(net: &Network) -> Self {
        net.feature_bias
    }

    pub fn add_feature(&mut self, feature_idx: usize, net: &Network) {
        for (i, d) in self.vals.iter_mut().zip(&net.feature_weights[feature_idx].vals) {
            *i += *d
        }
    }

    #[allow(dead_code)]
    pub fn remove_feature(&mut self, feature_idx: usize, net: &Network) {
        for (i, d) in self.vals.iter_mut().zip(&net.feature_weights[feature_idx].vals) {
            *i -= *d
        }
    }
}

// ---------------------------------------------------------------------
// FEN parsing and Chess768 feature encoding
// ---------------------------------------------------------------------

/// One piece on the board, in absolute (non-perspective-flipped) terms.
struct Piece {
    /// 0 = white, 1 = black
    colour: usize,
    /// 0=pawn, 1=knight, 2=bishop, 3=rook, 4=queen, 5=king
    kind: usize,
    /// 0..64, a1 = 0, h8 = 63
    square: usize,
}

struct Position {
    pieces: Vec<Piece>,
    /// 0 = white to move, 1 = black to move
    stm: usize,
}

fn parse_fen(fen: &str) -> Result<Position, String> {
    let parts: Vec<&str> = fen.split_whitespace().collect();
    let board_str = *parts.first().ok_or("empty FEN")?;
    let stm = match parts.get(1).copied() {
        Some("b") => 1,
        _ => 0, // treat missing/anything-else as white to move
    };

    const PIECE_CHARS: &str = "PNBRQKpnbrqk";

    let ranks: Vec<&str> = board_str.split('/').collect();
    if ranks.len() != 8 {
        return Err(format!("FEN board must have 8 ranks, found {}", ranks.len()));
    }

    let mut pieces = Vec::new();

    // FEN lists ranks from 8 down to 1; rank 1 is our rank index 0.
    for (i, rank_str) in ranks.iter().enumerate() {
        let rank = 7 - i;
        let mut file = 0usize;

        for ch in rank_str.chars() {
            if let Some(d) = ch.to_digit(10) {
                file += d as usize;
            } else {
                let idx = PIECE_CHARS
                    .find(ch)
                    .ok_or_else(|| format!("invalid piece character '{ch}'"))?;

                if file >= 8 {
                    return Err(format!("rank '{rank_str}' overflows 8 files"));
                }

                pieces.push(Piece {
                    colour: idx / 6,
                    kind: idx % 6,
                    square: rank * 8 + file,
                });
                file += 1;
            }
        }

        if file != 8 {
            return Err(format!("rank '{rank_str}' does not sum to 8 files"));
        }
    }

    Ok(Position { pieces, stm })
}

/// Builds the two perspective accumulators (us = side to move, them = other
/// side) for a position, using Chess768's feature layout:
/// `feature = colour * 384 + piece_kind * 64 + square`, mirroring the square
/// (^ 56) and flipping colour to get the opposite-perspective feature.
fn build_accumulators(net: &Network, pos: &Position) -> (Accumulator, Accumulator) {
    let mut us = Accumulator::new(net);
    let mut them = Accumulator::new(net);

    for p in &pos.pieces {
        let white_feat = p.colour * 384 + p.kind * 64 + p.square;
        let black_feat = (1 - p.colour) * 384 + p.kind * 64 + (p.square ^ 56);

        let (us_feat, them_feat) = if pos.stm == 0 {
            (white_feat, black_feat)
        } else {
            (black_feat, white_feat)
        };

        us.add_feature(us_feat, net);
        them.add_feature(them_feat, net);
    }

    (us, them)
}

/// Parses `fen`, builds accumulators, and runs the forward pass.
/// Returns the evaluation in centipawns, RELATIVE TO THE STM.
fn evaluate_fen(net: &Network, fen: &str) -> Result<i32, String> {
    let pos = parse_fen(fen)?;
    let (us, them) = build_accumulators(net, &pos);
    Ok(net.evaluate(&us, &them))
}

pub fn evaluate_fen_api(fen: &str) -> Result<i32, Box<dyn std::error::Error>>{
    match evaluate_fen(&NNUE, fen) {
        Ok(eval) => Ok(eval),
        Err(e) => Err(e.into()),
    }
}

// fn main() {
//     let args: Vec<String> = std::env::args().skip(1).collect();
//     if args.is_empty() {
//         eprintln!("Usage: nnue_eval \"<FEN>\"");
//         std::process::exit(1);
//     }
//     // Joined in case the shell splits an unquoted FEN on whitespace.
//     let fen = args.join(" ");

//     match evaluate_fen(&NNUE, &fen) {
//         Ok(eval) => println!("{eval}"),
//         Err(e) => {
//             eprintln!("Error: {e}");
//             std::process::exit(1);
//         }
//     }
// }